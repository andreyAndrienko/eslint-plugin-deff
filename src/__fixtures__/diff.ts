const diff = `diff --git a/fixme.js b/fixme.js
index 4886604..83c3014 100644
--- a/fixme.js
+++ b/fixme.js
@@ -1,0 +2,2 @@ if (new Date().getTime()) console.log("curly");
+if (new Date().getTime()) console.log("curly");
+if (new Date().getTime()) console.log("curly");`;

const staged = `diff --git a/fixme.js b/fixme.js
index 4886604..3238811 100644
--- a/fixme.js
+++ b/fixme.js
@@ -1,0 +2 @@ if (new Date().getTime()) console.log("curly");
+if (new Date().getTime()) console.log("curly");`;

const hunks = `diff --git a/dirty.js b/dirty.js
index 4d2637c..99dc494 100644
--- a/dirty.js
+++ b/dirty.js
@@ -1 +0 @@
-
@@ -3,7 +2,4 @@ function myFunc() {
-
-
-
-if (!myFunc(true == myVar))
-  myRes = false,
-  someotherThing = null
-  unrelated = true;
+  if (true == myVar)
+    if (true == myVar)
+      if (true == myVar) if (true == myVar) if (true == myVar) return;
+}
@@ -10,0 +7,2 @@ if (!myFunc(true == myVar))
+if (!myFunc(true == myVar)) (myRes = false), (someotherThing = null);
+unrelated = true;`;

const includingOnlyRemovals = `diff --git a/dirty.js b/dirty.js
index cb3c131..874b8f9 100644
--- a/dirty.js
+++ b/dirty.js
@@ -1,0 +2,2 @@ if (new Date().getTime()) console.log("curly");
+if (new Date().getTime()) console.log("curly");
+if (new Date().getTime()) console.log("curly");
@@ -17,2 +16,0 @@ import { a } from "../components/a";
-import { b } from "../context/b";
-import { c } from "../context/c";`;

const filenamesAB = `a/dirty.js
b/dirty.js
`;

const filenamesA = `a/dirty.js
`;

const diffFileList = "file1\nfile2\nfile3\n";

export {
  diff,
  staged,
  hunks,
  includingOnlyRemovals,
  filenamesAB,
  filenamesA,
  diffFileList,
};
