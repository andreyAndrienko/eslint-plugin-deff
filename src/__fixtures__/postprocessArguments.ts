import type { Linter } from "eslint";

const postprocessArguments = [
  [
    [
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 1,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
      },
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 2,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
      },
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 3,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
      },
    ],
  ],
  `/mock filename with quotes ", semicolons ; and spaces.js`,
] as [Linter.LintMessage[][], string];

export { postprocessArguments };
