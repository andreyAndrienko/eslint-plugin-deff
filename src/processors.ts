import type { Linter } from "eslint";
import type { Range } from "./git";
import {
  getDiffFileList,
  getDiffForFile,
  getIgnorePatterns,
  getRangesForDiff,
} from "./git";

const STAGED = true;

const isLineWithinRange = (line: number) => (range: Range) =>
  range.isWithinRange(line);

const diff = {
  postprocess: (
    messages: Linter.LintMessage[][],
    filename: string
  ): Linter.LintMessage[] => {
    const shouldKeepFile = getDiffFileList().includes(filename);

    return shouldKeepFile
      ? messages
          .map((message) =>
            message.filter(({ fatal, line }) => {
              const shouldKeepLine = getRangesForDiff(
                getDiffForFile(filename)
              ).some(isLineWithinRange(line));

              return fatal ?? shouldKeepLine;
            })
          )
          .reduce((a, b) => a.concat(b), [])
      : [];
  },

  supportsAutofix: true,
};

const diffConfig = {
  plugins: ["diff"],
  overrides: [
    {
      files: ["*"],
      processor: "diff/diff",
    },
  ],
  ignorePatterns: getIgnorePatterns(),
};

const staged = {
  postprocess: (
    messages: Linter.LintMessage[][],
    filename: string
  ): Linter.LintMessage[] => {
    const shouldKeepFile = getDiffFileList().includes(filename);

    return shouldKeepFile
      ? messages
          .map((message) =>
            message.filter(({ fatal, line }) => {
              const shouldKeepLine = getRangesForDiff(
                getDiffForFile(filename, STAGED)
              ).some(isLineWithinRange(line));

              return fatal ?? shouldKeepLine;
            })
          )
          .reduce((a, b) => a.concat(b), [])
      : [];
  },

  supportsAutofix: true,
};

const stagedConfig = {
  plugins: ["diff"],
  overrides: [
    {
      files: ["*"],
      processor: "diff/staged",
    },
  ],
  ignorePatterns: getIgnorePatterns(STAGED),
};

export { diff, diffConfig, staged, stagedConfig };
