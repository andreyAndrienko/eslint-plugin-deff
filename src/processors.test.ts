import * as child_process from "child_process";
import { mocked } from "ts-jest/utils";
import {
  diff as fixtureDiff,
  staged as fixtureStaged,
} from "./__fixtures__/diff";

jest.mock("child_process");
const mockedChildProcess = mocked(child_process, true);
mockedChildProcess.execSync.mockReturnValue(
  Buffer.from(`/mock filename with quotes ", semicolons ; and spaces.js`)
);

import { diff, diffConfig, staged, stagedConfig } from "./processors";
import { postprocessArguments } from "./__fixtures__/postprocessArguments";

const [messages, filename] = postprocessArguments;

describe("processors", () => {
  it("diff", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureDiff));

    expect(diff.postprocess(messages, filename)).toMatchSnapshot();

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
  });

  it("staged", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureStaged));

    expect(staged.postprocess(messages, filename)).toMatchSnapshot();

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
  });
});

describe("configs", () => {
  it("diff", () => {
    expect(diffConfig).toMatchSnapshot();
  });

  it("staged", () => {
    expect(stagedConfig).toMatchSnapshot();
  });
});
