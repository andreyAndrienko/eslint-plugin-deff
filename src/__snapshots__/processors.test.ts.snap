// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configs diff 1`] = `
Object {
  "ignorePatterns": Array [],
  "overrides": Array [
    Object {
      "files": Array [
        "*",
      ],
      "processor": "diff/diff",
    },
  ],
  "plugins": Array [
    "diff",
  ],
}
`;

exports[`configs staged 1`] = `
Object {
  "ignorePatterns": Array [],
  "overrides": Array [
    Object {
      "files": Array [
        "*",
      ],
      "processor": "diff/staged",
    },
  ],
  "plugins": Array [
    "diff",
  ],
}
`;

exports[`processors diff 1`] = `
Array [
  Object {
    "column": 1,
    "line": 2,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
  Object {
    "column": 1,
    "line": 3,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
]
`;

exports[`processors staged 1`] = `
Array [
  Object {
    "column": 1,
    "line": 2,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
]
`;
